File: src/wordsmith/util/strings.py
"""String utilities used by Wordsmith components."""

from __future__ import annotations

VOWELS = {"a", "e", "i", "o", "u"}


def starts_with_vowel(text: str) -> bool:
    """Return True if the first word starts with a vowel sound."""
    stripped = text.strip()
    word_chars: list[str] = []
    for char in stripped:
        if char.isalpha():
            word_chars.append(char)
        elif word_chars:
            break
    if not word_chars:
        return False

    word = "".join(word_chars).lower()
    silent_h_prefixes = ("honest", "honor", "honour", "hour", "heir", "herb")
    hard_u_prefixes = ("uni", "use", "user", "ufo", "euro", "one", "once")

    if word.startswith(silent_h_prefixes):
        return True
    if word.startswith(hard_u_prefixes):
        return False
    return word[0] in VOWELS


def first_upper(text: str) -> str:
    """Uppercase the first alphabetic character, leaving the rest untouched."""
    for index, char in enumerate(text):
        if char.isalpha():
            return f"{text[:index]}{char.upper()}{text[index + 1:]}"
    return text


def title_case(text: str) -> str:
    """Apply title casing with basic English small-word rules."""
    small_words = {
        "a",
        "an",
        "and",
        "as",
        "at",
        "but",
        "by",
        "for",
        "in",
        "nor",
        "of",
        "on",
        "or",
        "per",
        "so",
        "the",
        "to",
        "up",
        "via",
        "vs",
        "with",
        "yet",
    }
    punctuation = "\"'“”‘’()[]{}.,;:!?/"

    words = text.split(" ")
    last_index = len(words) - 1
    result: list[str] = []

    for index, word in enumerate(words):
        if word == "":
            continue

        leading = ""
        trailing = ""
        core = word

        while core and core[0] in punctuation:
            leading += core[0]
            core = core[1:]
        while core and core[-1] in punctuation:
            trailing = core[-1] + trailing
            core = core[:-1]

        if not core:
            result.append(word)
            continue

        core_lower = core.lower()
        if index not in {0, last_index} and core_lower in small_words:
            core_cased = core_lower
        else:
            core_cased = core_lower.capitalize()

        result.append(f"{leading}{core_cased}{trailing}")

    return " ".join(result)

File: packages/typescript/src/strings.ts
const LETTER = /\p{L}/u;
const CASED_LETTER = /[\p{Ll}\p{Lu}\p{Lt}]/u;
const VOWELS = new Set(["a", "e", "i", "o", "u"]);

export function startsWithVowel(text: string): boolean {
  const letters: string[] = [];
  for (const character of text.trim()) {
    if (LETTER.test(character)) {
      letters.push(character);
    } else if (letters.length > 0) {
      break;
    }
  }
  if (letters.length === 0) {
    return false;
  }

  const word = letters.join("").toLowerCase();
  if (
    ["honest", "honor", "honour", "hour", "heir", "herb"].some((prefix) =>
      word.startsWith(prefix),
    )
  ) {
    return true;
  }
  if (
    ["uni", "use", "user", "ufo", "euro", "one", "once"].some((prefix) =>
      word.startsWith(prefix),
    )
  ) {
    return false;
  }
  return VOWELS.has(word[0] ?? "");
}

export function firstUpper(text: string): string {
  const match = LETTER.exec(text);
  if (match?.index === undefined) {
    return text;
  }
  const character = match[0];
  return `${text.slice(0, match.index)}${character.toUpperCase()}${text.slice(match.index + character.length)}`;
}

export function capitalized(text: string): string {
  let afterUncased = true;
  let result = "";
  for (const character of text) {
    if (CASED_LETTER.test(character)) {
      result += afterUncased ? character.toUpperCase() : character.toLowerCase();
      afterUncased = false;
    } else {
      result += character;
      afterUncased = true;
    }
  }
  return result;
}

const SMALL_WORDS = new Set([
  "a",
  "an",
  "and",
  "as",
  "at",
  "but",
  "by",
  "for",
  "in",
  "nor",
  "of",
  "on",
  "or",
  "per",
  "so",
  "the",
  "to",
  "up",
  "via",
  "vs",
  "with",
  "yet",
]);
const PUNCTUATION = new Set([...`"'“”‘’()[]{}.,;:!?/`]);

export function titleCase(text: string): string {
  const words = text.split(" ");
  const lastIndex = words.length - 1;
  const result: string[] = [];

  for (const [index, word] of words.entries()) {
    if (word === "") {
      continue;
    }
    let core = word;
    let leading = "";
    let trailing = "";
    while (core.length > 0 && PUNCTUATION.has(core[0] as string)) {
      leading += core[0];
      core = core.slice(1);
    }
    while (core.length > 0 && PUNCTUATION.has(core.at(-1) as string)) {
      trailing = `${core.at(-1)}${trailing}`;
      core = core.slice(0, -1);
    }
    if (core === "") {
      result.push(word);
      continue;
    }

    const lower = core.toLowerCase();
    const transformed =
      index !== 0 && index !== lastIndex && SMALL_WORDS.has(lower)
        ? lower
        : firstUpper(lower);
    result.push(`${leading}${transformed}${trailing}`);
  }
  return result.join(" ");
}
